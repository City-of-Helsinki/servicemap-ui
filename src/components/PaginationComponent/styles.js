
// Styles
export default theme => ({
  arrowIcon: {
    fontSize: 18,
  },
  arrowFlip: {
    transform: 'scaleX(-1)',
  },
  borderBlack: {
    border: '1px solid #000000',
  },
  button: {
    margin: theme.spacing.unitHalf,
    height: 32,
    width: 32,
    minHeight: 32,
    minWidth: 32,
  },
  buttonContainer: {
    flexDirection: 'row',
    margin: `${theme.spacing.unit}px ${theme.spacing.unitDouble}px`,
    padding: theme.spacing.unitHalf,
  },
  list: {
    display: 'inherit',
    flexDirection: 'row',
    listStyleType: 'none',
    margin: 0,
    padding: 0,
  },
  listContainer: {
    flexDirection: 'row',
    margin: 0,
  },
  pageElement: {
    color: 'black',
    cursor: 'pointer',
    fontWeight: 'normal',
    textDecoration: 'none',
  },
  pageElementActive: {
    cursor: 'auto',
    fontWeight: 'normal',
    textDecoration: 'underline',
  },
  pageItem: {
    margin: 0,
    marginLeft: theme.spacing.unitHalf,
    marginRight: theme.spacing.unitHalf,
  },
});
