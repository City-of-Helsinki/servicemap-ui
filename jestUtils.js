import React from 'react';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { ThemeProvider } from '@mui/material/styles';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import themes from './src/themes';
import english from './src/i18n/en';

const mockStore = configureStore([]);

// Mock props for intl provider
const intlMock = {
  locale: 'en',
  messages: english,
  wrapRichTextChunksInFragment: false,
};

const Providers = (mockState) => ({ children }) => {
  const store = mockStore(mockState);
  return (
    <Provider store={store}>
      <IntlProvider {...intlMock}>
        <ThemeProvider theme={themes.SMTheme}>
          {children}
        </ThemeProvider>
      </IntlProvider>
    </Provider>
  );
}

export const getRenderWithProviders = mockState => component => render(component, { wrapper: Providers(mockState) });
