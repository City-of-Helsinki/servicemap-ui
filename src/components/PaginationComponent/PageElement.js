import React from 'react';
import PropTypes from 'prop-types';
import { intlShape } from 'react-intl';
import { Typography, withStyles, ButtonBase } from '@material-ui/core';
import styles from './styles';

// Page number element
const PageElement = ({
  classes, className, intl, isActive, number, onClick, ...rest
}) => {
  const newClassName = `${className} ${classes.pageItem} ${isActive ? classes.pageElementActive : classes.pageElement}`;
  return (
    <li>
      <ButtonBase
        role="link"
        disabled={isActive}
        onClick={onClick}
        tabIndex={isActive ? null : '0'}
      >
        <Typography
          variant="subtitle1"
          component="p"
          className={newClassName}
          {...rest}
        >
          <Typography variant="srOnly">
            {
            isActive
              ? intl.formatMessage({ id: 'general.pagination.currentlyOpenedPage' }, { count: number })
              : intl.formatMessage({ id: 'general.pagination.openPage' }, { count: number })
          }
          </Typography>
          <span aria-hidden="true">
            {number}
          </span>
        </Typography>
      </ButtonBase>
    </li>
  );
};

PageElement.propTypes = {
  classes: PropTypes.objectOf(PropTypes.any).isRequired,
  className: PropTypes.string,
  intl: intlShape.isRequired,
  isActive: PropTypes.bool.isRequired,
  number: PropTypes.number.isRequired,
  onClick: PropTypes.func.isRequired,
};

PageElement.defaultProps = {
  className: '',
};

export default withStyles(styles)(PageElement);
