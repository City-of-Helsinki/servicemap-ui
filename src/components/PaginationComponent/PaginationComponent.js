/* eslint-disable no-underscore-dangle */
import React from 'react';
import PropTypes from 'prop-types';
import { ArrowForwardIos } from '@material-ui/icons';
import Container from '../Container';
import PageElement from './PageElement';
import SMButton from '../ServiceMapButton';

class PaginationComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const {
      classes, current, handlePageChange, intl, maxShownPages, pageCount,
    } = this.props;

    // Return if only 1 page
    if (pageCount === 1) {
      return null;
    }

    const buttonClass = classes.borderBlack;

    const pages = [];
    const maxVisible = pageCount < maxShownPages ? pageCount : maxShownPages;
    const min = current > maxShownPages ? current - maxShownPages + 1 : 1;
    let max = current < maxVisible ? maxVisible : (current + 1);
    max = max > pageCount ? pageCount : max; // Don't allow anything above total page count

    for (let i = min; i <= max; i += 1) {
      pages.push(
        <PageElement
          key={i}
          number={i}
          intl={intl}
          isActive={current === i}
          onClick={() => { handlePageChange(i, pageCount); }}
        />,
      );
    }
    return (
      <Container className={classes.buttonContainer}>
        {
          // Button backwards one page
          <SMButton
            aria-label={intl.formatMessage({ id: 'general.pagination.previous' })}
            className={`${classes.button} ${classes.arrowFlip} ${buttonClass}`}
            onClick={(e) => {
              e.preventDefault();
              handlePageChange(current - 1, pageCount);
            }}
            disabled={current === 1}
            color={current !== 1 ? 'primary' : 'default'}
            variant="contained"
            role="link"
          >
            <ArrowForwardIos className={classes.arrowIcon} />
          </SMButton>
        }
        {
          // Button forward one page
          <SMButton
            aria-label={intl.formatMessage({ id: 'general.pagination.next' })}
            className={`${classes.button} ${buttonClass}`}
            onClick={(e) => {
              e.preventDefault();
              handlePageChange(current + 1, pageCount);
            }}
            disabled={current === pageCount}
            color={current !== pageCount ? 'primary' : 'default'}
            variant="contained"
            role="link"
          >
            <ArrowForwardIos className={classes.arrowIcon} />
          </SMButton>
        }
        {
          // Page numbers
          <Container className={classes.listContainer}>
            <ul className={classes.list}>
              {pages}
            </ul>
          </Container>
        }
      </Container>
    );
  }
}

PaginationComponent.propTypes = {
  classes: PropTypes.objectOf(PropTypes.any).isRequired,
  current: PropTypes.number.isRequired,
  handlePageChange: PropTypes.func.isRequired,
  intl: PropTypes.objectOf(PropTypes.any).isRequired,
  maxShownPages: PropTypes.number,
  pageCount: PropTypes.number.isRequired,
};

// Default props
PaginationComponent.defaultProps = {
  maxShownPages: 7,
};

export default PaginationComponent;
