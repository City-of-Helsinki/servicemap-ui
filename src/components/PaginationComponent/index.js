import { withStyles } from '@material-ui/core';
import { injectIntl } from 'react-intl';
import PaginationComponent from './PaginationComponent';
import styles from './styles';

export default withStyles(styles)(injectIntl(PaginationComponent));
