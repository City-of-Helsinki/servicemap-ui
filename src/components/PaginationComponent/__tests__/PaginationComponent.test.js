// Link.react.test.js
import React from 'react';
import { fireEvent } from '@testing-library/react';
import PaginationComponent from '../index';
import { getRenderWithProviders } from '../../../../jestUtils';
import englishTranslations from '../../../i18n/en';

// Mock props for intl provider
const intlMock = {
  locale: 'en',
  messages: {
    'general.pagination.previous': 'Aiempi sivu',
    'general.pagination.next': 'Seuraava sivu',
    'general.pagination.openPage': 'Avaa sivu {count}',
    'general.pagination.currentlyOpenedPage': 'Sivu {count}, avattu',
  },
  wrapRichTextChunksInFragment: false,
};

// Generic required props for ResultItem
const mockProps = {
  current: 2,
  handlePageChange: jest.fn(),
  maxShownPages: 5,
  pageCount: 8,
};

const renderWithProviders = getRenderWithProviders({});

describe('<PaginationComponent />', () => {
  it('should work', () => {
    const { container } = renderWithProviders(<PaginationComponent {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('simulates handlePageChange on click event', () => {
    const mockCallBack = jest.fn((newCurrent, totalCount) => ({ newCurrent, totalCount }));
    const { getAllByRole } = renderWithProviders(<PaginationComponent {...mockProps} handlePageChange={mockCallBack} />);

    // component.find('PageElement ForwardRef(ButtonBase)').at(2).simulate('click');
    fireEvent.click(getAllByRole('link')[1]);
    expect(mockCallBack.mock.calls.length).toEqual(1);

    // Expect handlePageChange to get first argument (newCurrent) correctly
    expect(mockCallBack.mock.results[0].value.newCurrent).toEqual(3);
    // Expect handlePageChange to get second argument (totalCount) correctly
    expect(mockCallBack.mock.results[0].value.totalCount).toEqual(mockProps.pageCount);
  });

  it('does set active correctly', () => {
    const { getAllByRole } = renderWithProviders(<PaginationComponent {...mockProps} />);
    expect(getAllByRole('link')[mockProps.current + 1]).toBeDisabled();
  });

  it('does use default accessibility attributes correctly', () => {
    const { getAllByRole } = renderWithProviders(<PaginationComponent {...mockProps} current={1} />);

    const buttons = getAllByRole('link');

    // Test previous page button accessibility
    expect(buttons[0]).toHaveAttribute('aria-label', englishTranslations['general.pagination.previous']);
    expect(buttons[0]).toBeDisabled();
    expect(buttons[0]).toHaveAttribute('role', 'link');
    // Test next page button accessibility
    expect(buttons[1]).toHaveAttribute('aria-label', englishTranslations['general.pagination.next']);
    expect(buttons[1]).not.toBeDisabled();
    expect(buttons[1]).toHaveAttribute('role', 'link');

    // Expect page 1 button to have opened text
    expect(buttons[2].querySelectorAll('p')[1]).toHaveTextContent('Page 1 currently opened');
    // expect page 2 button to have open new page text
    expect(buttons[3].querySelectorAll('p')[1]).toHaveTextContent('Open page 2');
  });
});
