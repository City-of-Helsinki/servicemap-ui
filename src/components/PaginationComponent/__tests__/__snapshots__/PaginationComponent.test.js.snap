// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaginationComponent /> should work 1`] = `
<div>
  <div
    class="Container-root-11    injectIntl(PaginationComponent)-buttonContainer-5"
  >
    <button
      aria-label="Previous page"
      class="MuiButtonBase-root SMButton SMButton-button-16  SMButton-marginRight-19 injectIntl(PaginationComponent)-button-4 injectIntl(PaginationComponent)-arrowFlip-2 injectIntl(PaginationComponent)-borderBlack-3 SMButton-primary-20 css-fvdnvy-MuiButtonBase-root"
      id="PaginationPreviousButton"
      role="link"
      tabindex="0"
      type="button"
      variant="contained"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium injectIntl(PaginationComponent)-arrowIcon-1 css-6yqi1c-MuiSvgIcon-root"
        data-testid="ArrowForwardIosIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
        />
      </svg>
    </button>
    <button
      aria-label="Next page"
      class="MuiButtonBase-root SMButton SMButton-button-16  SMButton-marginRight-19 injectIntl(PaginationComponent)-button-4 injectIntl(PaginationComponent)-borderBlack-3 SMButton-primary-20 css-fvdnvy-MuiButtonBase-root"
      id="PaginationNextButton"
      role="link"
      tabindex="0"
      type="button"
      variant="contained"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium injectIntl(PaginationComponent)-arrowIcon-1 css-6yqi1c-MuiSvgIcon-root"
        data-testid="ArrowForwardIosIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
        />
      </svg>
    </button>
    <div
      class="Container-root-11    injectIntl(PaginationComponent)-listContainer-7"
    >
      <ul
        class="injectIntl(PaginationComponent)-list-6"
      >
        <li>
          <button
            class="MuiButtonBase-root css-fvdnvy-MuiButtonBase-root"
            role="link"
            tabindex="0"
            type="button"
          >
            <p
              class="MuiTypography-root MuiTypography-subtitle1  PageElement-pageItem-33 PageElement-pageElement-31 css-dlbwt1-MuiTypography-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-1ak20dk-MuiTypography-root"
                style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
              >
                Open page 1
              </p>
              <span
                aria-hidden="true"
              >
                1
              </span>
            </p>
          </button>
        </li>
        <li>
          <button
            class="MuiButtonBase-root Mui-disabled css-fvdnvy-MuiButtonBase-root"
            disabled=""
            role="link"
            tabindex="-1"
            type="button"
          >
            <p
              class="MuiTypography-root MuiTypography-subtitle1  PageElement-pageItem-33 PageElement-pageElementActive-32 css-dlbwt1-MuiTypography-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-1ak20dk-MuiTypography-root"
                style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
              >
                Page 2 currently opened
              </p>
              <span
                aria-hidden="true"
              >
                2
              </span>
            </p>
          </button>
        </li>
        <li>
          <button
            class="MuiButtonBase-root css-fvdnvy-MuiButtonBase-root"
            role="link"
            tabindex="0"
            type="button"
          >
            <p
              class="MuiTypography-root MuiTypography-subtitle1  PageElement-pageItem-33 PageElement-pageElement-31 css-dlbwt1-MuiTypography-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-1ak20dk-MuiTypography-root"
                style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
              >
                Open page 3
              </p>
              <span
                aria-hidden="true"
              >
                3
              </span>
            </p>
          </button>
        </li>
        <li>
          <button
            class="MuiButtonBase-root css-fvdnvy-MuiButtonBase-root"
            role="link"
            tabindex="0"
            type="button"
          >
            <p
              class="MuiTypography-root MuiTypography-subtitle1  PageElement-pageItem-33 PageElement-pageElement-31 css-dlbwt1-MuiTypography-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-1ak20dk-MuiTypography-root"
                style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
              >
                Open page 4
              </p>
              <span
                aria-hidden="true"
              >
                4
              </span>
            </p>
          </button>
        </li>
        <li>
          <button
            class="MuiButtonBase-root css-fvdnvy-MuiButtonBase-root"
            role="link"
            tabindex="0"
            type="button"
          >
            <p
              class="MuiTypography-root MuiTypography-subtitle1  PageElement-pageItem-33 PageElement-pageElement-31 css-dlbwt1-MuiTypography-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-1ak20dk-MuiTypography-root"
                style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
              >
                Open page 5
              </p>
              <span
                aria-hidden="true"
              >
                5
              </span>
            </p>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
